registry/parser/parser.go

```diff
@@ -0,0 +1,178 @@
+package parser
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v4/registry"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/registry/exec"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/registry/state"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/scale"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+)
+
+type Event struct {
+	Name    string
+	Fields  map[string]any
+	EventID types.EventID
+	Phase   *types.Phase
+	Topics  []types.Hash
+}
+
+type EventParser interface {
+	GetEvents(blockHash types.Hash) ([]*Event, error)
+	ParseEvents(sd *types.StorageDataRaw) ([]*Event, error)
+}
+
+type eventParser struct {
+	stateProvider   state.Provider
+	registryFactory registry.Factory
+
+	eventStorageExecutor exec.RetryableExecutor[*types.StorageDataRaw]
+	eventParsingExecutor exec.RetryableExecutor[[]*Event]
+
+	eventRegistry registry.EventRegistry
+	meta          *types.Metadata
+}
+
+func NewParser(
```

Comment:
Not sure if we are going to have another type of parser, but for clarity maybe makes sense to rename this:
```suggestion
func NewEventParser(
```

Reply:
Yea, I forgot to change this since initially it was called parser, thank you!